components/nexus-cellar-hazelcast/src/main/java/org/sonatype/nexus/hazelcast/internal/HazelcastInstanceProvider.java

```diff
@@ -0,0 +1,171 @@
+ * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
+ *
+ * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
+ * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
+ * Eclipse Foundation. All other trademarks are the property of their respective owners.
+ */
+package org.sonatype.nexus.hazelcast.internal;
+
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+import java.util.Map;
+
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+
+import org.sonatype.nexus.common.app.ApplicationDirectories;
+import org.sonatype.nexus.hazelcast.HazelcastMemberAttributesCustomizer;
+import org.sonatype.sisu.goodies.common.ComponentSupport;
+
+import com.google.common.base.Throwables;
+import com.hazelcast.config.Config;
+import com.hazelcast.config.MemberAttributeConfig;
+import com.hazelcast.config.XmlConfigBuilder;
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.util.ExceptionUtil;
+import org.apache.karaf.cellar.core.utils.CombinedClassLoader;
+import org.eclipse.sisu.EagerSingleton;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceRegistration;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Expose the shared {@link HazelcastInstance}.
+ *
+ * @since 3.0
+ */
+@Named(HazelcastInstanceProvider.INSTANCE_NAME)
+@EagerSingleton
+public class HazelcastInstanceProvider
+    extends ComponentSupport
+    implements Provider<HazelcastInstance>
+{
+  public static final String INSTANCE_NAME = "nexus";
+
+  static {
+    // HACK: workaround https://github.com/orientechnologies/orientdb/pull/4609
+    ExceptionUtil.class.toString();
+  }
+
+  private final BundleContext bundleContext;
+
+  private final ApplicationDirectories directories;
+
+  private final List<HazelcastMemberAttributesCustomizer> memberAttributesCustomizers;
+
+  private final CombinedClassLoader combinedClassLoader;
+
+  private HazelcastInstance instance;
+
+  private ServiceRegistration<?> registration;
+
+  @Inject
+  public HazelcastInstanceProvider(final BundleContext bundleContext,
+                                   final ApplicationDirectories directories,
+                                   final List<HazelcastMemberAttributesCustomizer> memberAttributesCustomizers)
+  {
+    this.bundleContext = checkNotNull(bundleContext);
+    this.directories = checkNotNull(directories);
+    this.memberAttributesCustomizers = checkNotNull(memberAttributesCustomizers);
+    this.combinedClassLoader = new CombinedClassLoader();
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) new org.sonatype.nexus.hazelcast.internal.HazelcastInstanceProvider(BundleContext, ApplicationDirectories, List) creates a org.apache.karaf.cellar.core.utils.CombinedClassLoader classloader, which should be performed within a doPrivileged block [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ADP_CREATE_CLASSLOADER_INSIDE_DO_PRIVILEGED)